leg_detector/src/leg_detector_lib.cpp

```diff
@@ -0,0 +1,387 @@
+#include "leg_detector/leg_detector_lib.h"
+
+ LegDetector::LegDetector(const std::string& model_path) :
+    mask_count_(0),
+    feat_count_(0),
+    next_p_id_(0),
+    no_observation_timeout_s_(0.5),
+    max_second_leg_age_s_(2.0),
+    max_track_jump_m_(1.0),
+    max_meas_jump_m_(1.0),
+    leg_pair_separation_m_(1.0),
+    fixed_frame_("odom"),
+    kal_p_(4),
+    kal_q_(0.002),
+    kal_r_(10),
+    use_filter_(true),
+    connected_thresh_(0.06),
+    min_points_per_group(5),
+    leg_reliability_limit_(0.7)
+  {
+    forest = cv::ml::StatModel::load<cv::ml::RTrees>(model_path);
+    feat_count_ = forest->getVarCount();
+    printf("Loaded forest with %d features: %s\n", feat_count_, model_path.c_str());
+    feature_id_ = 0;
+  }
+
+
+  LegDetector::~LegDetector()
+  {
+  }
+
+  double LegDetector::distance(list<SavedFeature*>::iterator it1,  list<SavedFeature*>::iterator it2)
+  {
+    Stamped<Point> one = (*it1)->position_, two = (*it2)->position_;
+    double dx = one[0] - two[0], dy = one[1] - two[1], dz = one[2] - two[2];
+    return sqrt(dx * dx + dy * dy + dz * dz);
+  }
+
+  void LegDetector::pairLegs()
+  {
+    // Deal With legs that already have ids
+    list<SavedFeature*>::iterator begin = saved_features_.begin();
+    list<SavedFeature*>::iterator end = saved_features_.end();
+    list<SavedFeature*>::iterator leg1, leg2, best, it;
+
+    for (leg1 = begin; leg1 != end; ++leg1)
+    {
+      // If this leg has no id, skip
+      if ((*leg1)->object_id == "")
+        continue;
+
+      leg2 = end;
+      best = end;
+      double closest_dist = leg_pair_separation_m_;
+      for (it = begin; it != end; ++it)
+      {
+        if (it == leg1) continue;
+
+        if ((*it)->object_id == (*leg1)->object_id)
+        {
+          leg2 = it;
+          break;
+        }
+
+        if ((*it)->object_id != "")
+          continue;
+
+        double d = distance(it, leg1);
+        if (((*it)->getLifetime() <= max_second_leg_age_s_)
+            && (d < closest_dist))
```

Comment:
and here, if you use `distance2`, you can compare `d < closest_dist2`, where `closest_dist2 = closest_dist * closest_dist`